### DISCUSSION TITLE
Web-based configuration management

### DISCUSSION THREAD

**Discussion by kjk2010** [micro-nova/AmpliPi#86] [Ideas]
First of all, thanks for developing such an interesting product. 
Is there a plan to add a web front end for managing the config file via the API? If not, I'm interested in throwing something together and getting people's input.

I bring this up primarily because of the internetradio stream feature. I envision a page where users can quickly search multiple internet radio directories and add stations to their AmpliPis, and also edit and remove existing ones. A source management page like this could also be used to configure Pandora and Spotify settings.

Very curious on your plans and thoughts. Thanks

**Comment by linknum23** [comment]
Adding a config page is on the roadmap, but we have not started working on it yet.
Since configuration can get a little more complicated we were planning on doing this after migrating the AmpliPi webapp to more of a progressive web app styled framework such as react.

Migrating to a framework like react has several advantages:
- Easy to create matching Android and iOS apps. Notably an Android app would be useful because on Android MDNS can only be enabled in apps. This allows the app to work without specifying an ip address. Safari has MDNS enabled by default so this is not really an issue on the iPhone.
- Cleanup and organize the front end structure
- More responsive website

This is just our current idea of how to approach this. We could also just add a config page and slowly restructure the javascript. What do you think?
As for the design our initial idea is to put the settings/configuration page as the 5th tab under a gear icon. We are interested to see what kind of design you were thinking. Nothing is set in stone right now, so this is a perfect time to start putting something together.

Feel free to start working on a prototype on this, we would love to work with you on designing this.

**Reply by kjk2010** [reply]
I like your plan of using React. I haven't used it myself, but have been interested in digging in to it. As I find time, I'll put some ideas together into code and let you know.

**Reply by linknum23** [reply]
Sounds great.

**Reply by linknum23** [reply]
Might be a good starting point: https://medium.com/@toricpope/transform-a-react-app-into-a-progressive-web-app-pwa-dea336bd96e6

**Reply by rickballard** [reply]
On the future web-based config page, can we have an ability to both download and share our config files?  That way, our community could more easily help each other get started, making  more complicated tasks more accessible, like adding internet radio streams or home assistant integrations.

**Reply by linknum23** [reply]
You can already download the config file: **⚙️** -> **Configuration and Reset** ->  Click **Download**. Sharing we might consider in the furture.

**Comment by kjk2010** [comment]
I put in a PR for a simple stream management page: https://github.com/micro-nova/AmpliPi/pull/153. It uses the included JQuery and Bootstrap, and is very basic, but might be useful prior to migrating to a full framework.
